Error handling

BasileiosKal opened the pull request:
Implicitly handle errors form the utility sub-routines in the core operations + address point **B** from #246

christianpaquin reviewed (approved):
Improves readability

andrewwhitehead commented on draft-irtf-cfrg-bbs-signatures.md:
```diff
@@ -620,28 +634,27 @@ Deserialization:
+6.  D = P1 + Q_1 * domain + H_i1 * msg_i1 + ... + H_iR * msg_iR
+7.  C = Bbar * r2^ + Abar * r3^ + H_j1 * m^_j1 + ... + H_jU * m^_jU + D * c
+8.  cv = calculate_challenge(Abar, Bbar, C, (i1, ..., iR),
+                                      (msg_i1, ..., msg_iR), domain, ph)
```
Also here on the indentation

BasileiosKal replied:
Fixed!

andrewwhitehead commented on draft-irtf-cfrg-bbs-signatures.md:
```diff
@@ -724,33 +733,26 @@ Definitions:
                  18 bytes.
-- seed_len = ceil((ceil(log2(r)) + k)/8), where r and k are defined by
-                                          the ciphersuite.
+- expand_len, the expend_len value defined by the ciphersuite.
```
```suggestion
- expand_len, defined by the ciphersuite.
```

andrewwhitehead commented on draft-irtf-cfrg-bbs-signatures.md:
```diff
@@ -898,25 +893,28 @@ Inputs:
+ABORT if:
+
+1. R > 2^64 - 1 or R != length(msg_array)
+2. length(ph) > 2^64 - 1
+3. for i in i_array, i > 2^64 - 1
```
I believe this would be checked by `serialize`?

BasileiosKal replied:
Yes. Iv added these checks here to ABORT instead of returning INVALID (which we would have to do if we were relying on `serialize` for the checks).

That way we don't have to check the result of `calculate_challenge` and `calculate_domain` in the core operations.

BasileiosKal commented on draft-irtf-cfrg-bbs-signatures.md:
```diff
@@ -528,35 +538,39 @@ Deserialization:
+9.  Bbar = B * r1 - Abar * e
+10. C = Bbar * r2 + Abar * r3 + H_j1 * m~_j1 + ... + H_jU * m~_jU
+11. c = calculate_challenge(Abar, Bbar, C, (i1, ..., iR),
+                                      (msg_i1, ..., msg_iR), domain, ph)
```
Fixed!

BasileiosKal commented:
### Breaking Changes
1. In `hash_to_scalar` removed the `counter`
2. In `create_generators` step 3, changed `I2OSP(i, 4)` to `I2OSP(i, 8)`


("1." is because I removed the checks for 0 and "2."  is to align the max number of messages with the max number of generators)

tplooker commented:
Ready for merge, subject to resolving the conflicts.
